From f021a92280dd75327b113db73f6d2da77deba0da Mon Sep 17 00:00:00 2001
From: Gustaw Lewandowski <gustaw.lewandowski@intel.com>
Date: Wed, 21 Nov 2018 17:31:11 +0100
Subject: [PATCH 205/292] ASoC: Intel: Add support for imr_alloc flag

Add new flag imr_alloc to skl_machine_pdata for enable
imr buffer allocation for given platform needed for
presilicon validation on FPGA.

Change-Id: I64c5680be37c9cb1265877411be78b9049afebcf
Signed-off-by: Gustaw Lewandowski <gustaw.lewandowski@intel.com>
---
 sound/soc/intel/common/sst-dsp-priv.h |  1 +
 sound/soc/intel/skylake/cnl-sst.c     | 50 +++++++++++++++++++++++++++
 sound/soc/intel/skylake/skl.c         |  6 ++--
 sound/soc/intel/skylake/skl.h         |  1 +
 4 files changed, 56 insertions(+), 2 deletions(-)

diff --git a/sound/soc/intel/common/sst-dsp-priv.h b/sound/soc/intel/common/sst-dsp-priv.h
index 65197400cae6..7bdefe1b128c 100644
--- a/sound/soc/intel/common/sst-dsp-priv.h
+++ b/sound/soc/intel/common/sst-dsp-priv.h
@@ -359,6 +359,7 @@ struct sst_dsp {
 	u32 intr_status;
 	const struct firmware *fw;
 	struct snd_dma_buffer dmab;
+	struct snd_dma_buffer imr_buf;
 };
 
 /* Size optimised DRAM/IRAM memcpy */
diff --git a/sound/soc/intel/skylake/cnl-sst.c b/sound/soc/intel/skylake/cnl-sst.c
index 2044ce052a0f..dfe6be6d54f6 100644
--- a/sound/soc/intel/skylake/cnl-sst.c
+++ b/sound/soc/intel/skylake/cnl-sst.c
@@ -29,6 +29,7 @@
 #include <linux/device.h>
 #include <asm/set_memory.h>
 #include <asm/cacheflush.h>
+#include <sound/soc-acpi.h>
 
 #include "../common/sst-dsp.h"
 #include "../common/sst-dsp-priv.h"
@@ -37,6 +38,7 @@
 #include "skl-sst-dsp.h"
 #include "skl-sst-ipc.h"
 #include "skl-fwlog.h"
+#include "skl-topology.h"
 
 #define CNL_FW_ROM_INIT		0x1
 #define CNL_FW_INIT		0x5
@@ -63,10 +65,56 @@
 #define CNL_ADSP_FW_HDR_OFFSET	0x2000
 #define CNL_ROM_CTRL_DMA_ID	0x9
 
+#define CNL_IMR_MEMSIZE					0x400000
+#define CNL_IMR_PAGES	((CNL_IMR_MEMSIZE + PAGE_SIZE - 1) >> PAGE_SHIFT)
+#define HDA_ADSP_REG_ADSPCS_IMR_CACHED_TLB_START	0x100
+#define HDA_ADSP_REG_ADSPCS_IMR_UNCACHED_TLB_START	0x200
+#define HDA_ADSP_REG_ADSPCS_IMR_SIZE			0x8
+
+/* Needed for presilicon platform based on FPGA */
+static int cnl_alloc_imr(struct sst_dsp *ctx)
+{
+	if (ctx->dsp_ops.alloc_dma_buf(ctx->dev, &ctx->imr_buf,
+	     CNL_IMR_MEMSIZE) < 0) {
+		dev_err(ctx->dev, "Alloc imr buffer failed\n");
+		return -ENOMEM;
+	}
+
+	set_memory_uc((unsigned long)ctx->imr_buf.area, CNL_IMR_PAGES);
+	writeq(virt_to_phys(ctx->imr_buf.area) + 1,
+		 ctx->addr.shim + HDA_ADSP_REG_ADSPCS_IMR_CACHED_TLB_START);
+	writeq(virt_to_phys(ctx->imr_buf.area) + 1,
+		 ctx->addr.shim + HDA_ADSP_REG_ADSPCS_IMR_UNCACHED_TLB_START);
+
+	writel(CNL_IMR_MEMSIZE, ctx->addr.shim
+		+ HDA_ADSP_REG_ADSPCS_IMR_CACHED_TLB_START
+		+ HDA_ADSP_REG_ADSPCS_IMR_SIZE);
+	writel(CNL_IMR_MEMSIZE, ctx->addr.shim
+		+ HDA_ADSP_REG_ADSPCS_IMR_UNCACHED_TLB_START
+		+ HDA_ADSP_REG_ADSPCS_IMR_SIZE);
+
+	memset(ctx->imr_buf.area, 0, CNL_IMR_MEMSIZE);
+
+	return 0;
+}
+
+static inline void cnl_free_imr(struct sst_dsp *ctx)
+{
+	ctx->dsp_ops.free_dma_buf(ctx->dev, &ctx->imr_buf);
+}
+
 static int cnl_prepare_fw(struct sst_dsp *ctx, const void *fwdata, u32 fwsize)
 {
 
 	int ret, stream_tag;
+	struct skl *skl = get_skl_ctx(ctx->dev);
+	struct skl_machine_pdata *pdata = (struct skl_machine_pdata *)
+						skl->mach->pdata;
+	if (pdata && pdata->imr_alloc) {
+		ret = cnl_alloc_imr(ctx);
+		if (ret < 0)
+			return ret;
+	}
 
 	stream_tag = ctx->dsp_ops.prepare(ctx->dev, 0x40, fwsize, &ctx->dmab,
 						SNDRV_PCM_STREAM_PLAYBACK);
@@ -109,6 +157,8 @@ static int cnl_prepare_fw(struct sst_dsp *ctx, const void *fwdata, u32 fwsize)
 	ctx->dsp_ops.cleanup(ctx->dev, &ctx->dmab, stream_tag,
 						SNDRV_PCM_STREAM_PLAYBACK);
 	cnl_dsp_disable_core(ctx, SKL_DSP_CORE0_MASK);
+	if (pdata && pdata->imr_alloc)
+		cnl_free_imr(ctx);
 
 	return ret;
 }
diff --git a/sound/soc/intel/skylake/skl.c b/sound/soc/intel/skylake/skl.c
index 9441c73abad8..e87f51d6755e 100644
--- a/sound/soc/intel/skylake/skl.c
+++ b/sound/soc/intel/skylake/skl.c
@@ -560,7 +560,10 @@ static int skl_find_machine(struct skl *skl, void *driver_data)
 {
 	struct hdac_bus *bus = skl_to_bus(skl);
 	struct snd_soc_acpi_mach *mach = driver_data;
-	struct skl_machine_pdata *pdata;
+	struct skl_machine_pdata *pdata = mach->pdata;
+
+	if (pdata && pdata->imr_alloc)
+		goto out;
 
 	mach = snd_soc_acpi_find_machine(mach);
 	if (!mach) {
@@ -574,7 +577,6 @@ static int skl_find_machine(struct skl *skl, void *driver_data)
 
 	skl->mach = mach;
 	skl->fw_name = mach->fw_filename;
-	pdata = mach->pdata;
 
 	if (pdata) {
 		skl->use_tplg_pcm = pdata->use_tplg_pcm;
diff --git a/sound/soc/intel/skylake/skl.h b/sound/soc/intel/skylake/skl.h
index 66c4a5f23f64..f0e24328f58b 100644
--- a/sound/soc/intel/skylake/skl.h
+++ b/sound/soc/intel/skylake/skl.h
@@ -176,6 +176,7 @@ struct skl_dma_params {
 
 struct skl_machine_pdata {
 	bool use_tplg_pcm; /* use dais and dai links from topology */
+	bool imr_alloc;
 };
 
 struct skl_dsp_ops {
-- 
2.21.0

